src/composables/request.js

```diff
@@ -66,86 +63,77 @@ following options:
   if (data != null && data instanceof File && data.size > 100000000) {
-    alert.danger(this.$t('mixin.request.alert.fileSize', { name: data.name }));
+    alert.danger(i18n.t('mixin.request.alert.fileSize', { name: data.name }));
     return Promise.reject(new Error('file size exceeds limit'));
   }
 
-  if (this.awaitingResponse != null) this.awaitingResponse = true;
+  // eslint-disable-next-line no-param-reassign
+  awaitingResponse.value = true;
 
-  const initialRoute = this.$route;
+  // `router` may be `null` in testing.
+  const initialRoute = router != null ? router.currentRoute.value : null;
   return http.request(withAuth(axiosConfig, requestData.session.token))
     .catch(error => {
-      // this.$route seems to be defined even after the component has been
-      // unmounted.
-      if (this.$route !== initialRoute) throw new Error('route change');
+      if (router != null && router.currentRoute.value !== initialRoute)
+        throw new Error('route change');
 
       if (fulfillProblem != null && error.response != null &&
         isProblem(error.response.data) && fulfillProblem(error.response.data))
         return error.response;
 
-      if (this.awaitingResponse != null) this.awaitingResponse = false;
+      // eslint-disable-next-line no-param-reassign
+      awaitingResponse.value = false;
 
       logAxiosError(logger, error);
-      alert.danger(requestAlertMessage(this.$i18n, error, problemToAlert));
+      alert.danger(requestAlertMessage(i18n, error, problemToAlert));
       throw error;
     })
     .then(response => {
-      if (this.$route !== initialRoute) throw new Error('route change');
-      if (this.awaitingResponse != null) this.awaitingResponse = false;
+      if (router != null && router.currentRoute.value !== initialRoute)
+        throw new Error('route change');
+      // eslint-disable-next-line no-param-reassign
+      awaitingResponse.value = false;
 
       return response;
     });
-}
-
-// @vue/component
-const mixin = {
-  inject: ['container'],
-  watch: {
-    $route() {
-      if (this.awaitingResponse != null) this.awaitingResponse = false;
-    }
-  },
-  methods: {
-    request,
-    post(url, data = undefined, config = undefined) {
-      const full = { ...config, method: 'POST', url };
-      if (data != null) full.data = data;
-      return this.request(full);
-    },
-    put(url, data, config = undefined) {
-      return this.request({ ...config, method: 'PUT', url, data });
-    },
-    patch(url, data, config = undefined) {
-      return this.request({ ...config, method: 'PATCH', url, data });
-    },
-    delete(url, config = undefined) {
-      return this.request({ ...config, method: 'DELETE', url });
-    }
-  }
 };
 
-export default () => mixin;
+export default () => {
+  const container = inject('container');
+  const awaitingResponse = ref(false);
+  const request = _request(container, awaitingResponse);
```

Comment:
can we use `.apply` here and use update value of awaitingResponse using `this`? that why we don't have to disable eslint above. Not totally sure if that will work

Reply:
Good question, I was thinking about something like this as well. I think something like what you describe would be possible. But it sort of feels like mostly a way to bypass ESLint. The same mutation would be happening, just via `this`.

Another strategy I considered was to pass a function to `_request()` to set `awaitingResponse.value`. `_request()` only ever needs to set `awaitingResponse.value`, not get it. So we could do something like:

```
const setAwaitingResponse = (value) => { awaitingResponse.value = value; };
const request = _request(container, setAwaitingResponse);
```

ESLint wouldn't object in that case. But I don't think that code is really clearer.

In general, I think this is going to come up sometimes with the Composition API. Within the Composition API, refs will be passed from function to function. And the way to change a ref is to set its `value` property, which ESLint will sometimes object to. But my sense is that that's a normal pattern within the Composition API. If something that creates a ref wants to prevent it from being set, it can call `readonly()` on it. That's what happens here: `awaitingResponse.value` is set internally within the composable, but when the composable returns it to the component, it calls `readonly()` on it first.

Reply:
make sense